### DISCUSSION TITLE
(best way to) estimate event size

### DISCUSSION THREAD

**Discussion by srstrickland** [vectordotdev/vector#19781] [Q&A]
Recently we had a rogue process start spamming megabyte log lines, and needless to say bad things started happening.  We installed a quick fix at the agent level to just drop anything whose `.message` field exceeds some threshold (via strlen), and it got us out of the woods.  Maybe in practice, checking the `message` field is sufficient, since all log messages start there no matter the source.  But we have a distributed topology, and don't control _all_ the agents (configs) in our ecosystem, so I need something similar enforced at the "central" layer (which receives data from all the various agents).  And technically at this point, it's possible that some upstream agent has already parsed a gigantic message into other fields, so it may be insufficient to check only the message field.

The simplest way I've found to estimate payload size is to just encode as json and take a strlen.  Obviously this is an overshot, but since a lot of data is serialized as JSON on the way out, it's not unreasonable.  But adding a full serialization just to count the bytes feels like overkill, and I'm wondering if there's a better way.  I couldn't find anything in the VRL docs, so at the moment this is all I have:

```
if (strlen(encode_json(.)) > some_threshold)  # could use length() too
```

Is there another option?  I figure since vector is doing a lot of monitoring around bytes moving through the system, this might be an easy opportunity to expose that information via a function like `estimate_size(<any>)`.  It feels like this would be far more efficient than allocating & using memory just for a string to be counted.

Thanks in advance for any pointers!

**Comment by jszwedko** [comment]
Hey!

That would definitely be one way to do it though, as you note, it would be expensive.

There is actually a trait on events in Vector that exposes a function to estimate the in-memory size of the event:

https://github.com/vectordotdev/vector/blob/0dce77620fbc240f6e880c6f49f7ef7f8bb5e3df/lib/vector-common/src/byte_size_of.rs#L12-L32

As well as another trait to estimate the encoded JSON size of the event:

https://github.com/vectordotdev/vector/blob/0dce77620fbc240f6e880c6f49f7ef7f8bb5e3df/lib/vector-core/src/event/estimated_json_encoded_size_of.rs#L43-L45

I think these would just need to be exposed in VRL by adding these functions to the `vrl::Target`  trait and then functions added to the VRL stdlib to access them. I think we'd be happy to see a PR for this if you are interested.
